torch/csrc/api/src/nn/init.cpp

```diff
@@ -123,7 +123,7 @@ Tensor orthogonal_(Tensor tensor, double gain) {
       "Only tensors with 2 or more dimensions are supported");
 
   const auto rows = tensor.size(0);
-  const auto columns = tensor.size(1);
+  const auto columns = tensor[0].numel();
```

Comment:
I'm a bit doubtful about the change. They're semantically the same, but the previous version is faster.

Reply:
They're only the same for 2 dimensional tensors. For tensors of 3 dimensions or more `numel()` counts all the elements in the sub-dimensions, whereas `size(1)` only counts the size of the 1st dimension.

For example: Calling `size(1)` on a [3, 4, 5, 6] shaped tensor returns 4, but calling `tensor[0].numel()` returns 120.

Reply:
Looking at the code below, this is what I think is correct:
```
const auto rows = tensor.size(-2);
const auto columns = tensor.size(-1);
```

A matrix of size (rows x columns) is created whose QR decomposition is performed and then is copied to the input tensor. We might have to expand and copy the tensor for tensors of dim > 2.

Reply:
Yeah, that would avoid an unnecessary potential copy. I can change it to do that instead. Should we change the [Python implementation](https://github.com/pytorch/pytorch/blob/90b8552c9811ee6e9bbc9dff1f1df30c00b663dd/torch/nn/init.py#L348) too?

Reply:
Yeah, that'd be great.